a10networks/a10-octavia
# [STACK-2612/2611] Automated licensing of vThunder-Amphora

## Description
[STACK-2611] - Adds the ability to configure primary & secondary DNS servers of vThunder-Amphora
[STACK-2612] - Adds automated licensing of vThunder-Amphora via flexpool tokens

Currently, the vThunder-as-Amphora usage mode is not production ready in large part due to the fact that the vThunder instances are not automatically licensed. The code in this PR is capable of licensing the devices though it is largely WIP.

### Bug
[STACK-2757](https://a10networks.atlassian.net/browse/STACK-2757)

Severity: High

The `_create_active_standby_amp` subflow called `GetVThunderByLB` and `vThunderComputeConnectivityWait` twice for the vMaster instead of once for the vMaster and the vBlade. This likely resulted in the errors observed in [STACK board](https://a10networks.atlassian.net/browse/STACK-2295)/PR #376. Those changes have not been reverted though as this has not been fully confirmed.

## Linked PR
https://github.com/a10networks/acos-client/pull/357/files

## Limitations
This only works with Capacity Flexpool and Capacity Flexpool burst subscriptions. Other licensing methods such as "Pay-as-You-Go" will not be supported.

## Jira Ticket
- [STACK-2611](https://a10networks.atlassian.net/browse/STACK-2611)
- [STACK-2612](https://a10networks.atlassian.net/browse/STACK-2612)
- [STACK-2757](https://a10networks.atlassian.net/browse/STACK-2757)

## Technical Approach
- Required: Overview of approach taken to solve the problem
- Optional: List of tasks completed w/ links 

## Config Changes
- Added `[glm_license]` as a section header
- Added `flexpool_token`
- Added `primary_dns`
- Added `secondary_dns`
- Added `allocate_bandwidth`

\<The below do not have an effect yet\>

- Added `proxy_host`
- Added `proxy_username`
- Added `proxy_password`
- Added `proxy_port`

\<end\>

- Changed help text across various options to read "vThunder-Amphora" instead of "vThunder".  This is to make a distinction between the common deployment of vThunder instances on Openstack and Amphora created via Octavia.
- Added `amp_mgmt_net` to better identify the management network in the configuration file and ensure it is slotted into the mgmt interface on the ACOS device. If it is not specified then the first network defined in `amp_boot_network_list` is used.
- Added `amp_license_network` which is used to specify the network to connect to the GLM from. If this is not used, then the `amp_mgmt_net` or first network in `amp_boot_network_list ` is used.

## Flavor Changes
{'dns': {'primary-dns': <primary_ip>, 'secondary-dns': <secondary_ip>}}

## Test Cases
N/A

## Manual Testing
Please see tests attached to stories in the comments.

[STACK-2611]: https://a10networks.atlassian.net/browse/STACK-2611
[STACK-2612]: https://a10networks.atlassian.net/browse/STACK-2612
